cpp/src/aztec3/circuits/rollup/merge/native_merge_rollup_circuit.cpp

```diff
@@ -116,18 +121,22 @@ std::array<fr, 2> compute_calldata_hash(std::array<abis::PreviousRollupData<NT>,
 
 // asserts that the end snapshot of previous_rollup 0 equals the start snapshot of previous_rollup 1 (i.e. ensure they
 // follow on from one-another). Ensures that right uses the tres that was updated by left.
-void assert_prev_rollups_follow_on_from_each_other(BaseOrMergeRollupPublicInputs left,
+void assert_prev_rollups_follow_on_from_each_other(DummyComposer& composer,
+                                                   BaseOrMergeRollupPublicInputs left,
                                                    BaseOrMergeRollupPublicInputs right)
 {
-    assert(left.end_private_data_tree_snapshot == right.start_private_data_tree_snapshot);
-    assert(left.end_nullifier_tree_snapshot == right.start_nullifier_tree_snapshot);
-    assert(left.end_contract_tree_snapshot == right.start_contract_tree_snapshot);
-    // void variables since despite using in assert, it says, "unused variable"
+    composer.do_assert(left.end_private_data_tree_snapshot == right.start_private_data_tree_snapshot,
+                       "input proofs have different private data tree snapshots");
+    composer.do_assert(left.end_nullifier_tree_snapshot == right.start_nullifier_tree_snapshot,
+                       "input proofs have different nullifier tree snapshots");
+    composer.do_assert(left.end_contract_tree_snapshot == right.start_contract_tree_snapshot,
+                       "input proofs have different contract tree snapshots");
+    // void variables since despite using in composer.do_assert, it says, "unused variable"
```

Comment:
Oh dear! That's surprising...

Reply:
Hmmm... that doesn't sound right.... It says that left/right are unused? That makes it sound like the do_assert is getting optimized away by the compiler?

Reply:
nvm I tried again with a clean build and it worked this time 🤦